wandb/sdk/artifacts/storage_handlers/fsspec_file_handler.py

```diff
@@ -0,0 +1,146 @@
+        self._scheme = scheme or "oss"
+        self._cache = get_artifact_file_cache()
+
+    def can_handle(self, parsed_url: "ParseResult") -> bool:
+        return parsed_url.scheme in self._scheme
+
+    def load_path(
+        self,
+        manifest_entry: ArtifactManifestEntry,
+        local: bool = False,
+    ) -> Union[URIStr, FilePathStr]:
+        """Load the file in the specified artifact given its corresponding entry.
+
+        Download the referenced artifact; return the path to the caller.
+
+        Arguments:
+            manifest_entry (ArtifactManifestEntry): The index entry to load
+
+        Returns:
+            (os.PathLike): A path to the file represented by `index_entry`
+        """
+        if not local:
+            assert manifest_entry.ref is not None
+            return manifest_entry.ref
+
+        if manifest_entry.ref is None:
+            raise ValueError(f"Cannot add path with no ref: {manifest_entry.path}")
+        fs, fs_path = fsspec.core.url_to_fs(str(manifest_entry.ref))
+
+        if not fs.exists(fs_path):
+            raise ValueError(
+                "fsspec file reference: Failed to find file at path %s" % fs_path
+            )
+
+        path, hit, cache_open = self._cache.check_md5_obj_path(
+            B64MD5(manifest_entry.digest),  # TODO(spencerpearson): unsafe cast
+            manifest_entry.size if manifest_entry.size is not None else 0,
+        )
+        if hit:
+            return path
+
+        md5 = fs.checksum(fs_path)
+        if md5 != manifest_entry.digest:
+            raise ValueError(
+                f"fsspec file reference: Digest mismatch for path {manifest_entry.path}: expected {manifest_entry.digest} but found {md5}"
+            )
+
+        filesystem.mkdir_exists_ok(os.path.dirname(path))
+        fs.get_file(fs_path, path)
+
+        return path
+
+    def store_path(
+        self,
+        artifact: "Artifact",
+        path: Union[URIStr, FilePathStr],
+        name: Optional[StrPath] = None,
+        checksum: bool = True,
+        max_objects: Optional[int] = None,
+    ) -> Sequence[ArtifactManifestEntry]:
+        max_objects = max_objects or DEFAULT_MAX_OBJECTS
+        # We have a single file or directory
+        entries = []
+        fs, fs_path = fsspec.core.url_to_fs(path)
+
+        def md5(path: str) -> Any:
+            return (
+                fs.checksum(path)
+                if checksum
+                else md5_string(str(fs.info(path)["size"]))
+            )
+
+        if fs.isdir(fs_path):
+            i = 0
+            start_time = time.time()
+            if checksum:
+                termlog(
+                    'Generating checksum for up to %i files in "%s"...\n'
+                    % (max_objects, fs_path),
+                    newline=False,
+                )
+            for root, _, files in fs.walk(fs_path):
+                for sub_path in files:
+                    i += 1
+                    if i > max_objects:
```

Comment:
Nit: we don't need `i`:
```suggestion
                    if len(entries) == max_objects:
```

Reply:
Also adjusted this in `fsspec_file_handler.py`